src/coreclr/src/jit/emitxarch.cpp

```diff
@@ -2669,6 +2669,45 @@ void emitter::emitLoopAlign()
     emitCurIGsize += 15;
 }
 
+/*****************************************************************************
+ *
+ *  The next instruction will be a loop head entry point
+ *  So insert a dummy instruction here to ensure that
+ *  the x86 I-cache alignment rule is followed.
+ */
+
+void emitter::emitVariableLoopAlign(unsigned short alignmentBoundary)
+{
+    unsigned short nPaddingBytes    = alignmentBoundary - 1;
+    unsigned short nAlignInstr      = (nPaddingBytes + (15 - 1)) / 15;
+    unsigned short instrDescSize    = nAlignInstr * SMALL_IDSC_SIZE;
+    unsigned short insAlignCount    = nPaddingBytes / 15;
+    unsigned short lastInsAlignSize = nPaddingBytes % 15;
+
+    // Ensure that all align instructions fall in same IG.
```

Comment:
Why is this required? I think you should avoid this, and deal with overflow groups. E.g., we have stress modes that allow for very small instruction groups.

Reply:
I hit an issue in PMI run when I realized that this is needed.

For non-adaptive case, imagine to have an align boundary of 32 bytes. We will be generating 3 `INS_align` instructions for such cases - 15 bytes, 15 bytes, 1 byte. Now, when generating these 3 instructions, if they all don't fit in current IG (because of `SC_IG_BUFFER_SIZE` limit), we will end up creating new IG and have some `INS_align` instructions in new IG. After that, we will end up marking the `IGF_ALIGN_LOOP` on the newly created IG [here](https://github.com/dotnet/runtime/blob/fa309cb74a406b97ceb49e26d8954c00818095c6/src/coreclr/src/jit/codegenlinear.cpp#L788) (which is wrong). Going forward, when we try to see if we can remove some of the `align` instructions, we end up doing the IG size checks and updates on wrong IG and hit [this assert](https://github.com/dotnet/runtime/blob/fa309cb74a406b97ceb49e26d8954c00818095c6/src/coreclr/src/jit/emit.cpp#L4537).